File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RxErrorHandler.java
package com.habitrpg.android.habitica.helpers;

import android.util.Log;

import com.habitrpg.android.habitica.BuildConfig;
import com.habitrpg.android.habitica.proxy.CrashlyticsProxy;

import java.io.EOFException;
import java.io.IOException;

import io.reactivex.functions.Consumer;
import okhttp3.internal.http2.ConnectionShutdownException;
import retrofit2.HttpException;

public class RxErrorHandler {

    static private RxErrorHandler instance;
    private CrashlyticsProxy crashlyticsProxy;

    public static void init(CrashlyticsProxy crashlyticsProxy) {
        instance = new RxErrorHandler();
        instance.crashlyticsProxy = crashlyticsProxy;

        //RxJavaHooks.setOnError(handleEmptyError());
    }

    public static Consumer<Throwable> handleEmptyError() {
        //Can't be turned into a lambda, because it then doesn't work for some reason.
        return RxErrorHandler::reportError;
    }

    public static void reportError(Throwable throwable) {
        if (BuildConfig.DEBUG) {
            try {
                Log.e("ObservableError", Log.getStackTraceString(throwable));
            } catch (Exception ignored) {}
        } else {
            if (!IOException.class.isAssignableFrom(throwable.getClass())
                    && !HttpException.class.isAssignableFrom(throwable.getClass())
                    && !retrofit2.HttpException.class.isAssignableFrom(throwable.getClass())
                    && !EOFException.class.isAssignableFrom(throwable.getClass())
            && !ConnectionShutdownException.class.isAssignableFrom(throwable.getClass())) {
                instance.crashlyticsProxy.logException(throwable);
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RxErrorHandler.kt
package com.habitrpg.android.habitica.helpers

import android.util.Log
import com.habitrpg.android.habitica.BuildConfig
import com.habitrpg.android.habitica.proxy.AnalyticsManager
import io.reactivex.rxjava3.functions.Consumer
import okhttp3.internal.http2.ConnectionShutdownException
import retrofit2.HttpException
import java.io.EOFException
import java.io.IOException

class RxErrorHandler {
    private var analyticsManager: AnalyticsManager? = null

    companion object {

        private var instance: RxErrorHandler? = null

        fun init(analyticsManager: AnalyticsManager) {
            instance = RxErrorHandler()
            instance?.analyticsManager = analyticsManager
        }

        fun handleEmptyError(): Consumer<Throwable> {
            //Can't be turned into a lambda, because it then doesn't work for some reason.
            return Consumer { reportError(it) }
        }

        fun reportError(throwable: Throwable) {
            if (BuildConfig.DEBUG) {
                try {
                    Log.e("ObservableError", Log.getStackTraceString(throwable))
                } catch (ignored: Exception) {
                }

            } else {
                if (!IOException::class.java.isAssignableFrom(throwable.javaClass)
                        && !HttpException::class.java.isAssignableFrom(throwable.javaClass)
                        && !retrofit2.HttpException::class.java.isAssignableFrom(throwable.javaClass)
                        && !EOFException::class.java.isAssignableFrom(throwable.javaClass)
                        && throwable !is ConnectionShutdownException) {
                    instance?.analyticsManager?.logException(throwable)
                }
            }
        }
    }
}
